### ISSUE TITLE
`empty_count` rule triggers on local symbols

### ISSUE THREAD

**Issue by pronebird** [realm/SwiftLint#5326]
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

`empty_count` violation is triggered on extension over `IteratorProtocol`

Sample code:

```swift
extension IteratorProtocol {
    mutating func skip(_ count: Int) {
        guard count > 0 else { return } // ERROR: Empty Count Violation: Prefer checking `isEmpty` over comparing `count` to zero (empty_count)

        for _ in 0..<count {
            _ = next()
        }
    }
}
```

##### Complete output when running SwiftLint, including the stack trace and command used

```bash
$ swiftlint lint
```

### Environment

* SwiftLint version (run `swiftlint version` to be sure)?
* Installation method used (Homebrew, CocoaPods, building from source, etc)?
* Paste your configuration file:

```yml
# insert yaml contents here
```

* Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)?
  If so, paste their relative paths and respective contents.
* Which Xcode version are you using (check `xcodebuild -version`)?
* Do you have a sample that shows the issue? Run `echo "[string here]" | swiftlint lint --no-cache --use-stdin --enable-all-rules`
  to quickly test if your example is really demonstrating the issue. If your example is more
  complex, you can use `swiftlint lint --path [file here] --no-cache --enable-all-rules`.

**Comment by SimplyDanny** [comment]
While this is indeed a bug, it's not something that can be fixed cleanly. SwiftLint relies only on syntax information to be very fast. Therefore, it doesn't know in which context `isEmpty` is really a valid suggestion. The only way to improve the situation is by adding a distinct Analyzer rule as a typesafe version of `empty_count`.

**Comment by rgoldberg** [comment]
@SimplyDanny Is this the same issue as the spurious warning I get for:

```swift
import Foundation

func countTest() {
    let count: Int = Date().hashValue % 10
    if count > 0 { // error: Empty Count Violation: Prefer checking `isEmpty` over comparing `count` to zero (empty_count)
        print("Count is greater than zero")
    } else {
        print("Count is zero or less")
    }
}
```

If so, you're saying that the syntax info doesn't record the difference between a local let and a computed var struct member, right? To a layman, it seems like it shouldn't be hard to record such differences.

Could swiftlint annotate any local let, var or parameter as a let, var, or  parameter, then when rules like `empty_count` run, they can not warn if the name in question is an unqualified variable name in the current list of locals? Obviously, that would require scoping to blocks, but would detecting `}` outside of comments & strings, and having a `[Set<String>]` to hold variable names per block, be that difficult or slow?

**Comment by SimplyDanny** [comment]
@rgoldberg: These are very good and valid questions!

In general, `count` could come from anywhere and the rule just assumes that if there exists `count` then there must be `isEmpty` as well. This leads to false positives from time to time like in the example code.

Now, you are right that if `count` is a local variable, the rule could know that and ignore these cases. We have a few rules that do that and it is conceivable to improve `empty_count` likewise.

The rule even has an option called `only_after_dot` that works around that. Implementing identifier tracking would make this option obsolete for a few more cases.

@pronebird: The option could also be of interest for you.
